cou/steps/hypervisor.py

```diff
@@ -148,29 +150,51 @@ def azs(self) -> AZs:
 
         return azs
 
-    def _generate_pre_upgrade_plan(self) -> PreUpgradeStep:
+    def _generate_pre_upgrade_steps(
+        self, target: OpenStackRelease, group: HypervisorGroup
+    ) -> list[PreUpgradeStep]:
         """Generate pre upgrade plan for all applications.
 
-        This section should create a plan with steps like changing charm config option, etc.
+        This section should create a list of steps like changing charm config option, etc.
 
-        :return: Pre-upgrade step with all needed pre-upgrade steps.
-        :rtype: PreUpgradeStep
+        :param target: OpenStack codename to upgrade.
+        :type target: OpenStackRelease
+        :param group: HypervisorGroup object
+        :type group: HypervisorGroup
+        :return: List of pre-upgrade steps.
+        :rtype: list[PreUpgradeStep]
         """
-        raise NotImplementedError
+        steps = []
+        for app in self.apps:
+            if app.name not in group.app_units:
+                logger.debug(
+                    "skipping application %s because it is not part of group %s",
+                    app.name,
+                    group.name,
+                )
+                continue
 
-    def _generate_hypervisor_group_upgrade_plan(
-        self, target: OpenStackRelease, group: HypervisorGroup
-    ) -> HypervisorUpgradePlan:
+            units = group.app_units[app.name]
+            logger.info("generating pre-upgrade steps for %s app and %s units", app.name, units)
+            steps.extend(app.pre_upgrade_steps(target, units))
+
+        return steps
+
+    def _generate_upgrade_steps(
+        self, target: OpenStackRelease, force: bool, group: HypervisorGroup
+    ) -> list[UpgradeStep]:
```

Comment:
I forgot the reason why having `force` is needed for upgrade steps. Is it to further filter machine if there are vms spinned up after plan stage but before the actual upgrade operation? If so, why we don't want to also check it in the pre-upgrade stage?

Reply:
To be able to run steps with force, yes for example for nova-comput it means that we can upgrade non empty hypervisors. 

> If so, why we don't want to also check it in the pre-upgrade stage?

I do not understand, why would you need it in pre-upgrade stage?

Reply:
> To be able to run steps with force, yes for example for nova-comput it means that we can upgrade non empty hypervisors.

Maybe I'm missing something. My understanding is that if `force` is recorded, we will not filter out machines that are non-empty, so upgrade steps will generate steps for them even without accepting a `force` parameter.

But if it is really needed, I believe we should also add it for pre-upgrade stage because to not by default upgrading packages if there are vms running. Unless that's not risky at all?

Reply:
Yes, that is one part. But we are passing force to the upgrade plan to not checking if empty hypervisor again (ignoring it).
We are checking it during planning and also part of upgrade phase, as upgrade step (e.g. [line](https://github.com/canonical/charmed-openstack-upgrader/blob/70f3814f31821dc0abef67048038fdbb33e59409/tests/unit/steps/test_hypervisor.py#L215)).